src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs

```diff
@@ -28,7 +28,11 @@ public static class TagHelperBlockRewriter
             var attributes = GetTagAttributes(tagName, validStructure, tag, descriptors, errorSink);
             var selfClosing = IsSelfClosing(tag);
 
-            return new TagHelperBlockBuilder(tagName, selfClosing, start, attributes, descriptors);
+            // All descriptors for an element should share a common prefix, grab the first.
+            var prefix = descriptors.First().Prefix;
+            var tagHelperTagName = tagName.Substring(prefix.Length);
```

Comment:
to confirm: when all associated tag helpers no-op, the prefix will _not_ appear in the generated HTML.  if so, fine, that sounds consistent with the opt-out (`!`) handling.

but the prefix isn't used later.  need only the short name when generating the `TagHelperScopeManager.Begin()` call.  so passing it around together with descriptors containing exactly the same `string` is not worthwhile.

what may be useful is calculating the global prefix length once per parse, say at the top of this class.  then to get the short name from the full matched text, only need the `Substring` call.